src/runtime.rs

```diff
@@ -0,0 +1,371 @@
+                            let rt = tokio::runtime::Builder::new_current_thread()
+                                .max_blocking_threads(1)
+                                .thread_name(blocking_name)
+                                .build()
+                                .expect("failed to build tokio runtime");
+                            rt.block_on(local.run_until(async move {
+                                while let Ok(Task(f)) = task_receiver.recv_async().await {
+                                    tokio::task::spawn_local(f());
+                                }
+                            }));
+                            tracing::trace!("runtime {} dropped", main_name);
+                        })
+                        .expect("failed to spawn OS thread")
+                })
+                .collect::<Vec<_>>();
+            Runtime {
+                handle: Handle {
+                    sender: Arc::new(Mutex::new(Some(task_sender))),
+                },
+                name,
+                handles,
+            }
+        }
+
+        pub fn handle(&self) -> &Handle {
+            &self.handle
+        }
+
+        /// Allow the given duration for a graceful shutdown, then abandon
+        /// the threads and return.
+        pub fn shutdown_timeout(mut self, duration: Duration) {
+            tracing::trace!("shutting down runtime with timeout");
+            // send the shutdown signal by dropping the sender
+            self.handle.sender.lock().unwrap().take();
+            // wait for all threads to finish
+            let start = std::time::Instant::now();
+            while start.elapsed() < duration {
+                if self.handles.iter().all(|h| h.is_finished()) {
+                    break;
+                }
+                std::thread::sleep(std::time::Duration::from_millis(10));
+            }
+            self.handles.clear();
+        }
+
+        /// Shutdown the runtime entirely in the background
+        pub fn shutdown_background(self) {
+            self.shutdown_timeout(Duration::from_secs(0));
+        }
+    }
+
+    /// Signal to the runtime that it should shutdown, and wait indefinitely
+    /// for it to do so.
+    impl Drop for Runtime {
+        fn drop(&mut self) {
+            tracing::trace!("dropping sender");
+            self.handle.sender.lock().unwrap().take();
+            tracing::trace!("waiting for runtimes to terminate");
+            for handle in self.handles.drain(..) {
+                handle.join().ok();
+            }
+        }
+    }
+
+    impl Handle {
+        fn sender(&self) -> flume::Sender<Task> {
+            let inner = self.sender.lock().unwrap();
+            let sender = inner.as_ref().unwrap().clone();
+            sender
+        }
+
+        pub async fn spawn<F, Fut>(&self, f: F)
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+        {
+            let f = || f().map(|_| ()).boxed_local();
```

Comment:
ufff

Reply:
You don't like the boxing? Or mapping away the result? Not much you can do about the boxing, and mapping away the result should be a noop for things that already return unit and very cheap otherwise.

Reply:
it’s just not very pretty, I like simple code..

Reply:
You can write this in a different way, but I am not sure if it will be simpler. E.g.

```rust
let f = || async { f(); }.boxed_local();
```

Reply:
I think I like that last option more.